include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_scan.h

```diff
@@ -178,10 +251,7 @@ single_pass_scan_impl(sycl::queue __queue, _InRange&& __in_rng, _OutRange&& __ou
             // Obtain unique ID for this work-group that will be used in decoupled lookback
             if (group.leader())
             {
-                sycl::atomic_ref<::std::uint32_t, sycl::memory_order::relaxed, sycl::memory_scope::device,
-                                 sycl::access::address_space::global_space>
-                    idx_atomic(status_flags[status_flags_elems - 1]);
-                tile_id_lacc[0] = idx_atomic.fetch_add(1);
+                tile_id_lacc[0] = scan_memory<_Type>::fetch_add_counter(num_wgs, scratch);
```

Comment:
The requirement to pass in `num_wgs` here is kinda ugly. Maybe the `memory_holder` thing I linked can help get rid of this. It's not a big deal either way, I guess. Arguably, the `dynamic_tile_id` functionality could be separated from the `scan_memory` class entirely.

Reply:
Agreed. I've created a small wrapper around the `dynamic_tile_id`. It's now a bit clunky because the counter is in the middle of the scratch, but it removes the need for calculating an offset all the time by storing the actual location as a "separate" entity. Let me know your thoughts.

Reply:
Comment outdated due to the refactor of how memory is managed.